extensions/scarb-doc/src/docs_generation.rs

```diff
@@ -1 +1,156 @@
-mod markdown;
+#![allow(dead_code)]
+use crate::types::{
+    Constant, Crate, Enum, ExternFunction, ExternType, FreeFunction, Impl, ImplAlias, ImplConstant,
+    ImplFunction, ImplType, Member, Module, Struct, Trait, TraitConstant, TraitFunction, TraitType,
+    TypeAlias, Variant,
+};
+
+pub mod markdown;
+
+#[derive(Default)]
+struct TopLevelItems<'a> {
+    pub modules: Vec<&'a Module>,
+    pub constants: Vec<&'a Constant>,
+    pub free_functions: Vec<&'a FreeFunction>,
+    pub structs: Vec<&'a Struct>,
+    pub enums: Vec<&'a Enum>,
+    pub type_aliases: Vec<&'a TypeAlias>,
+    pub impl_aliases: Vec<&'a ImplAlias>,
+    pub traits: Vec<&'a Trait>,
+    pub impls: Vec<&'a Impl>,
+    pub extern_types: Vec<&'a ExternType>,
+    pub extern_functions: Vec<&'a ExternFunction>,
+}
+
+fn collect_all_top_level_items(crate_: &Crate) -> TopLevelItems {
+    let mut top_level_items = TopLevelItems::default();
+
+    top_level_items.modules.push(&crate_.root_module);
+
+    collect_all_top_level_items_internal(&mut top_level_items, &crate_.root_module);
+    top_level_items
+}
+
+fn collect_all_top_level_items_internal<'a, 'b>(
```

Comment:
Wouldn't writing this iteratively instead of recursion make this much simpler?

Reply:
I don't see how - modules contain submodules, so it is more readable as recursive function imo. It just requires proficiency in lifetimes ;p